dvc/remote/base.py

```diff
@@ -497,22 +508,43 @@ def _cache_is_copy(self, path_info):
         self.cache_type_confirmed = True
         return self.cache_types[0] == "copy"
 
-    def _save_dir(self, path_info, checksum, save_link=True):
-        cache_info = self.checksum_to_path_info(checksum)
-        dir_info = self.get_dir_cache(checksum)
+    def _save_dir(self, path_info, checksum, save_link=True, tree=None):
+        if tree:
+            checksum = self._save_tree(path_info, tree)
+        else:
+            dir_info = self.get_dir_cache(checksum)
 
-        for entry in Tqdm(
-            dir_info, desc="Saving " + path_info.name, unit="file"
-        ):
-            entry_info = path_info / entry[self.PARAM_RELPATH]
-            entry_checksum = entry[self.PARAM_CHECKSUM]
-            self._save_file(entry_info, entry_checksum, save_link=False)
+            for entry in Tqdm(
+                dir_info, desc="Saving " + path_info.name, unit="file"
+            ):
+                entry_info = path_info / entry[self.PARAM_RELPATH]
+                entry_checksum = entry[self.PARAM_CHECKSUM]
+                self._save_file(entry_info, entry_checksum, save_link=False)
 
-        if save_link:
-            self.state.save_link(path_info)
+            if save_link:
+                self.state.save_link(path_info)
 
+        cache_info = self.checksum_to_path_info(checksum)
         self.state.save(cache_info, checksum)
-        self.state.save(path_info, checksum)
+        if not tree or is_working_tree(tree):
+            self.state.save(path_info, checksum)
+
+    def _save_tree(self, path_info, tree):
+        # save tree directory to cache, collect dir cache during walk and
+        # return the resulting dir checksum
+        dir_info = []
+        for fname in tree.walk_files(path_info):
+            checksum = tree.get_file_checksum(fname)
+            file_info = {
+                self.PARAM_CHECKSUM: checksum,
+                self.PARAM_RELPATH: fname.relative_to(path_info).as_posix(),
+            }
+            self._save_file(fname, checksum, tree=tree)
+            dir_info.append(file_info)
```

Comment:
The actual behavior for this `walk` depends on how `tree` was initialized.

Git files will always be saved directly from the tree in `_save_file()` via `tree.open()`
For DVC outs:
* If `fetch=False` and `stream=False`, `walk` will not recurse into DVC out directories
* If `fetch=True`, DVC outs will be fetched when `walk` recurses into the out directory
    * `_save_file` will skip the `tree.open()`->copy for fetched outs (since they will already exist in cache)
* If `stream=True`, DVC outs will be saved by streaming in `_save_file()` via `tree.open()`

Basically, for erepo's and import/get, we probably always want the behavior from `fetch=True`.